### PR TITLE
feat: add a lot of new code

### PR DESCRIPTION
**Pull request by EvgeniErmakov**

### CODE REVIEW

File: .idea/compiler.xml [hunk 1]
```diff
@@ -1,6 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Comment (jisqaqov) [inline .idea/compiler.xml:1]: please delete the IDE files, include in gitignore

File: build.gradle [hunk 2]
```diff
@@ -1,19 +1,35 @@
...
+allprojects {
+    plugins.withType(JavaPlugin) {
+        dependencies {
+            implementation 'org.postgresql:postgresql:42.3.1'
+            compileOnly group: 'org.projectlombok', name: 'lombok', version: '1.18.20'
+            annotationProcessor 'org.projectlombok:lombok:1.18.20'
+            implementation group: 'org.springframework.boot', name: 'spring-boot-starter-hateoas', version: '2.5.1'
+            testImplementation group: 'org.springframework.boot', name: 'spring-boot-starter-test', version: '2.5.1'
+            implementation group: 'org.springframework.boot', name: 'spring-boot-starter-validation', version: '2.5.1'
+            implementation group: 'org.springframework.boot', name: 'spring-boot-starter-data-jpa', version: '2.5.1'
```
Comment (jisqaqov) [inline build.gradle:20]: spring data jpa using is resticted, it should be deleted, use hibernate instead

File: src/main/java/com/epam/esm/controller/CertificateController.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    /**
+     * @param id
+     * @param patchDTO
+     * @return
+     */
+    @PatchMapping(value = "/{id}")
+    public CertificateDTO patch(@PathVariable @Min(MIN_ID) Long id, @Valid @RequestBody PatchDTO patchDTO) {
+        certificateService.applyPatch(id, patchDTO);
```
Comment (jisqaqov) [inline src/main/java/com/epam/esm/controller/CertificateController.java]: let applyPath to return the certificate

File: src/main/java/com/epam/esm/model/exception/UserNotFoundException.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
+package com.epam.esm.model.exception;
+
+public class UserNotFoundException extends CustomServiceException {
```
Comment (jisqaqov) [inline src/main/java/com/epam/esm/model/exception/UserNotFoundException.java]: please create common EntityNotFoundException

File: src/main/java/com/epam/esm/model/exception/UserNotFoundException.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+    private static final int USER_NOT_FOUND_STATUS_CODE = 404;
+    private static final String USER_NOT_FOUND_MESSAGE = "Error! User has been not found!";
+
+    public UserNotFoundException(String message) {
+        super(message);
```
Comment (jisqaqov) [inline src/main/java/com/epam/esm/model/exception/UserNotFoundException.java]: where the message fields is used?

File: .idea/misc.xml [hunk 6]
```diff
@@ -1,10 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Comment (ShiniTada) [inline .idea/misc.xml:1]: please don't commit .idea files

File: src/main/java/com/epam/esm/model/entity/ErrorResponse.java [hunk 7]
```diff
@@ -0,0 +1,16 @@
+package com.epam.esm.model.entity;
+
```
Comment (jisqaqov) [inline src/main/java/com/epam/esm/model/entity/ErrorResponse.java]: can you please move this class to controller package

File: src/main/java/com/epam/esm/model/exception/OrderNotFoundException.java [hunk 8]
```diff
@@ -0,0 +1,21 @@
+package com.epam.esm.model.exception;
+
+public class OrderNotFoundException extends CustomServiceException {
```
Comment (jisqaqov) [inline src/main/java/com/epam/esm/model/exception/OrderNotFoundException.java]: can you create general exception EntityNotFoundException

File: src/main/java/com/epam/esm/model/exception/TagNotFoundException.java [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+    private static final int TAG_NOT_FOUND_STATUS_CODE = 404;
+    private static final String TAG_NOT_FOUND_MESSAGE = "Error! Tag has been not found!";
+
+    public TagNotFoundException(String message) {
+        super(message);
+    }
+
+    @Override
+    public int getErrorCode() {
```
Comment (jisqaqov) [inline src/main/java/com/epam/esm/model/exception/TagNotFoundException.java]: unnecessary method, please delete

File: src/main/java/com/epam/esm/model/entity/QuerySpecification.java [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import java.math.BigDecimal;
+import java.util.List;
+
+@Data
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
```
Comment (jisqaqov) [inline src/main/java/com/epam/esm/model/entity/QuerySpecification.java]: there already a class ParameterSpecificationDTO, this class need to be deleted

File: .gitignore [hunk 11]
```diff
@@ -0,0 +1,7 @@
+/.gradle/
+/.idea/
+gradlew.bat
```
Comment (jisqaqov) [inline .gitignore:3]: add idea classes to ignore, gradle classes are needed, do not ignore them

File: src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+    @Override
+    @Transactional
+    public CertificateDTO update(CertificateDTO updateDTO) {
+        Optional<Certificate> optional = certificateDAO.findById(updateDTO.getId());
+        if (optional.isEmpty() || !optional.get().isActive()) {
+            throw new CertificateNotFoundException(updateDTO.getId().toString());
+        }
+        Certificate certificate = optional.get();
+        Certificate update = mapperDTO.convertDTOToCertificate(updateDTO);
+        certificate = certificateDAO.update(certificate, update);
```
Comment (jisqaqov) [inline src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java]: update logic should be in the service layer, please move certificateDAO.update(certificate, update) from DAO layer to service

File: src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+        if (!certificate.isActive()) {
+            throw new CertificateNotFoundException(id.toString());
+        }
+        Certificate update = mapperDTO.convertPatchDTOToCertificate(patchDTO);
+        certificate = certificateDAO.applyPatch(certificate, update);
+        return mapperDTO.convertCertificateToDTO(certificate);
+    }
+
+    @Override
+    public List<CertificateDTO> findAllByOrderId(Long id, Page page) {
```
Comment (jisqaqov) [inline src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java:103]: here the pagination is not needed

File: src/main/java/com/epam/esm/service/impl/OrderServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+    @Override
+    @Transactional
+    public OrderDTO create(OrderDTO orderDTO) {
+        orderDTO.setCost(new BigDecimal(0));
+        List<Certificate> certificates = new ArrayList<>();
+        Order order = mapperDTO.convertDTOToOrder(orderDTO);
+        User user = userDAO.findById(orderDTO.getUserId())
+            .orElseThrow(() -> new UserNotFoundException(orderDTO.getUserId().toString()));
+        order.setUser(user);
+        orderDTO.getCertificateId().forEach(id -> {
```
Comment (jisqaqov) [inline src/main/java/com/epam/esm/service/impl/OrderServiceImpl.java:64]: too big peace of code, split the method. add getCertificates method to retrieve order certificates from DB

File: src/main/java/com/epam/esm/service/impl/OrderServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+        order.setCertificates(certificates);
+        order = orderDAO.create(order);
+        return mapperDTO.convertOrderToDTO(order);
+    }
+
+    @Override
+    public void delete(Long id) {}
+
+    @Override
+    public List<OrderDTO> findAllOrdersByUserId(Long id, Page page) {
```
Comment (jisqaqov) [inline src/main/java/com/epam/esm/service/impl/OrderServiceImpl.java:80]: no pagination needed here

File: src/main/java/com/epam/esm/dao/TagDAO.java [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+
+import com.epam.esm.model.entity.Tag;
+
+import java.util.Optional;
+
+public interface TagDAO extends CommonDAO<Tag, Long> {
+
+    Optional<Tag> findByName(String name);
+
+    Tag findOrCreate(Tag tag);
```
Comment (jisqaqov) [inline src/main/java/com/epam/esm/dao/TagDAO.java]: bad practice, do use only create or find methods

**Review by jisqaqov — CHANGES_REQUESTED** [review]

File: src/main/java/com/epam/esm/service/impl/OrderServiceImpl.java (hunk 15 above)
Comment (EvgeniErmakov) [reply src/main/java/com/epam/esm/service/impl/OrderServiceImpl.java:77]: A person can have a lot of orders.
The role of a person can be performed by a company or other legal entity. I suggest leaving pagination.

File: src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java (hunk 13 above)
Comment (EvgeniErmakov) [reply src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java:103]: A person can have a lot of orders.
The role of a person can be performed by a company or other legal entity. I suggest leaving pagination.

Comment (EvgeniErmakov) [reply src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java:103]: > 

A person can have a lot of orders.
The role of a person can be performed by a company or other legal entity. I suggest leaving pagination.

File: src/main/java/com/epam/esm/service/impl/OrderServiceImpl.java (hunk 15 above)
Comment (EvgeniErmakov) [reply src/main/java/com/epam/esm/service/impl/OrderServiceImpl.java:80]: > 

A person can have a lot of orders.
The role of a person can be performed by a company or other legal entity. I suggest leaving pagination.

File: src/main/java/com/epam/esm/model/exception/UserNotFoundException.java (hunk 5 above)
Comment (EvgeniErmakov) [reply src/main/java/com/epam/esm/model/exception/UserNotFoundException.java]: > 
When I search for a user and cannot find him in the database. In this case, I throw an exception.

File: src/main/java/com/epam/esm/controller/CertificateController.java (hunk 3 above)
Comment (EvgeniErmakov) [reply src/main/java/com/epam/esm/controller/CertificateController.java]: > 
I am trying to communicate between microservices or with the user using DTO. I see no reason to convey the entity to him.

File: .idea/compiler.xml (hunk 1 above)
Comment (EvgeniErmakov) [reply .idea/compiler.xml:1]: I added all unnecessary folders and settings to the git ignore

**Review by jisqaqov — APPROVED** [review]

**Review by jisqaqov — APPROVED** [review]

**Pull request merged**